Demo "edit" transform that supports various quite specific edits

wlupton opened the pull request:
...of the statement tree

It has the same functionality as an earlier (and now-withdrawn) proposed "yang-edit" output format (PR #246).

wlupton commented:
@mbj4668 are you inclined to merge this or something like it (the name could change)? Or would you prefer that it remain outside the core pyang code base? Tx, W.

mbj4668 commented on pyang/transforms/edit.py:
```diff
@@ -0,0 +1,338 @@
+    return child
+
+
+# XXX is there a proper function for this?
```
no

wlupton replied:
I deleted the comment.

mbj4668 commented:
I think we can include this plugin.  Please fix the conflicts and then I'll merge it.

wlupton commented on pyang/transforms/edit.py:
```diff
@@ -0,0 +1,338 @@
+    return summary if summary else 'TBD'
+
+
+# XXX need proper insertion in canonical order; currently just appending
```
I didn't want to investigate, make (and test...) this change right now, so I updated the comment to reference yang.py and its use of grammar.sort_canonical(). Hope that's OK for now.

wlupton commented:
I have found a problem when testing this transform against some real YANG. The YANG (actually it's [this YANG](https://github.com/BroadbandForum/yang/blob/master/standard/networking/bbf-qos-traffic-mngt.yang)) ends like this:
```
  augment '/if:interfaces/if:interface' {
    description
      "Augments the interface with traffic management data.";

    uses tm-root;
  }
```
...and I noticed that, when running the transform with the YANG output format, I get these warnings:
```
bbf-qos-traffic-mngt.yang:7: warning: imported module ietf-interfaces not used
bbf-qos-traffic-mngt.yang:11: warning: imported module bbf-yang-types not used
bbf-qos-traffic-mngt.yang:15: warning: imported module bbf-qos-types not used
```
The generated YANG does contain the expected statements. However, when using the tree output format, I get the same warnings but the tree does not contain the expected statements (the augment is empty):
```
  augment /if:interfaces/if:interface:
```
I am wondering whether some additional validation needs to be done after invoking the transform. The current logic (in the pyang binary in master) is as follows:
```
    # transform modules
    if len(xform_objs) > 0 and len(modules) > 0:
        for xform_obj in xform_objs:
            try:
                if not xform_obj.transform(ctx, modules):
                    for module in modules:
                        statements.validate_module(ctx, module)
            except error.TransformError as e:
                if e.msg != "":
                    sys.stderr.write(e.msg + '\n')
                sys.exit(e.exit_code)
```
The edit transform's transform() method doesn't return a value (so it should return None, yes?) so I believe that all the modules should be re-validated.

What's going on here?

wlupton commented:
@mbj4668, any thoughts on my earlier comment / question? I thought perhaps you might immediately know what's going on, but if not I could investigate further. Perhaps (given that this is just a demo transform) it would be OK to merge this PR anyway, and track this as an open issue against master.

mbj4668 commented:
No I don't know what's going on.

When you say "demo" transform, it makes me wonder... should we move it to a new "exampes" directory?  Or should we keep it where it is, but clearly document it as being a "demo" transform?

wlupton commented:
OK, I'll take a look.

"demo" is perhaps not quite the right word. It's definitely functional, but perhaps a bit quirky and maybe not sufficiently general to be the basis for the one-and-only "edit" transform.

Moving it to an "examples" directory sounds good. My only slight concern is that it would require an explicit `--plugindir` to find it. Might you consider adding a `pyang` `--include-examples` option (or something like that) to automatically include plugins (might be transforms or output formats) found in the distribution examples directory?

wlupton commented:
I looked into this warning. First I narrowed it down to this simple YANG:
```
module test {
  namespace "urn:test";
  prefix test;
  import ietf-interfaces {
    prefix if;
  }
  augment '/if:interfaces/if:interface' {
    container foo {
    }
  }
}
```
Then I added debug statements before and after the post-transform `validate_module()` call (of course this was after having narrowed the problem down to here!):
```
    # transform modules
    if len(xform_objs) > 0 and len(modules) > 0:
        for xform_obj in xform_objs:
            try:
                if not xform_obj.transform(ctx, modules):
                    for module in modules:
                        sys.stderr.write('before validate %r\n' % ctx.errors)
                        statements.validate_module(ctx, module)
                        sys.stderr.write('after validate %r\n' % ctx.errors)
            except error.TransformError as e:
                if e.msg != "":
                    sys.stderr.write(e.msg + '\n')
                sys.exit(e.exit_code)
```
Now running `pyang` with and without the `edit` transform gives the following:
```
% pyang -t edit test.yang
before validate []
after validate [(<pyang.error.Position object at 0x106c584c8>, 'UNUSED_IMPORT', 'ietf-interfaces')]
test.yang:4: warning: imported module ietf-interfaces not used
```
```
% pyang test.yang
```
I initially assumed that the warning must be triggered by the way in which the `edit` transform manipulates the statement tree but then I realised that the warning is generated even if the transform does nothing, and returns immediately!

The initial call to `validate_module()` was from `add_parsed_module()`. It seem like you simply can't expect to call it twice?

Therefore my question is what (if anything) I should be doing about validating modules after they've been transformed. I believe that I've already determined that _something_ has to be done because (for example) the tree format uses `i_children` so it's necessary to recalculate that.

Finally, any thoughts on the idea adding something like an `--include-examples` option?

mbj4668 commented:
This is tricky.  The problem is that some of the v_* functions keep state, so in this case it is v_expand_2_augment():

    if hasattr(stmt, 'i_target_node'):
        # already expanded
        return

To do this properly we'd have to reset all state before calling validate_module the second time.  But this needs to be done in a "safe" way.  

I have done one attempt in mbj/transform-reset-env.  Please check it out.

wlupton commented:
Thanks. I tried this with the `edit` transform in a real-life case and there does seem to be a problem. 

I haven't (yet) tried to narrow it down too much, but you should be able to see the same thing if you run from the top-level of a `YangModels/yang` checkout (including checking out the `standard/bbf` submodule).

On the `BroadbandForum:feature/add-demo-edit-transform` pyang branch:
```
% pyang --path standard/ietf/RFC/ --path standard/bbf/standard --max-line-length=70 --keep-comments --transform=edit --format=yang --edit-yang-version=1.1 --yang-line-length=70 --edit-previous-revision-date=2018-07-13 --edit-revision-date=2018-12-03 standard/bbf/standard/networking/bbf-qos-policies.yang >bbf-qos-policies.yang.edited
standard/bbf/standard/networking/bbf-qos-policies.yang:19: warning: imported module bbf-yang-types not used
```
With the new changes merged into the above pyang branch:
```
% pyang --path standard/ietf/RFC/ --path standard/bbf/standard --max-line-length=70 --keep-comments --transform=edit --format=yang --edit-yang-version=1.1 --yang-line-length=70 --edit-previous-revision-date=2018-07-13 --edit-revision-date=2018-12-03 standard/bbf/standard/networking/bbf-qos-policies.yang >bbf-qos-policies.yang.edited
standard/bbf/standard/networking/bbf-qos-policies.yang:215: error: type "qos-policy-profile-ref" not found in module bbf-qos-policies
standard/bbf/standard/networking/bbf-qos-policies.yang:222: error: type "qos-policy-profile-ref" not found in module bbf-qos-policies
```
Please let me know if you'd like me to try to narrow it down further (e.g. simpler YANG or simpler command line).

wlupton commented:
I wondered whether this would make any difference (resetting all the modules before re-adding any):
```
                if not xform_obj.transform(ctx, modules):
                    for module in modules:
                        module.internal_reset()
                    for module in modules:
                        ctx.add_parsed_module(module)
```
...but it didn't.

wlupton commented:
@mbj4668, any thoughts on this? I'm happy to be given homework, e.g. trying to narrow down the problem.

mbj4668 commented:
@wlupton, yes I really want to solve this as well.  I am writing a new 'sanitize' plugin, which fits perfect as a transform.  So if you want to have a look, it would be great.

wlupton commented:
@mbj4668, if we moved this to an 'examples' directory and added an `--include-examples` option (or similar; not vital) would you be OK with merging this and creating a new issue for the warnings?

wlupton commented:
@mbj4668, I finally got around to looking at this again and can only assume that I had failed to use the changes in your mbj/transform-reset-env branch. Sorry about that. I tried it again and now get a different error (with this same module):
```
bbf-qos-policies.yang:250: error: type "qos-policy-profile-ref" not found in module bbf-qos-policies
bbf-qos-policies.yang:257: error: type "qos-policy-profile-ref" not found in module bbf-qos-policies
```

I looked into this and believe that when the module is parsed for the second time (after the transform) something is going wrong with the statement hierarchy. I added some debug output to search_typedef() and see the following:
```
type.search_typedef(qos-policy-profile-ref)
  type qos-policy-profile-ref
  leaf ingress-qos-policy-profile
  augment /if:interfaces/if:interface
  module bbf-qos-policies
  FOUND
type.search_typedef(qos-policy-profile-ref)
  type qos-policy-profile-ref
  leaf egress-qos-policy-profile
  augment /if:interfaces/if:interface
  module bbf-qos-policies
  FOUND
bbf-qos-policies.internal_reset()
type.search_typedef(qos-policy-profile-ref)
  type qos-policy-profile-ref
  leaf ingress-qos-policy-profile
  list interface
  container interfaces
  module ietf-interfaces
type.search_typedef(qos-policy-profile-ref)
  type qos-policy-profile-ref
  leaf egress-qos-policy-profile
  list interface
  container interfaces
  module ietf-interfaces
```

I can dig deeper... but do you have any immediate thoughts?

Here's the modified search_typedef().
```
def search_typedef(stmt, name):
    """Search for a typedef in scope
    First search the hierarchy, then the module and its submodules."""
    w = sys.stderr.write if name in ['qos-policy-profile-ref'] \
        else lambda x: None
    w('%s.search_typedef(%s)\n' % (stmt.keyword, name))
    mod = stmt.i_orig_module
    while stmt is not None:
        w('  %s %s\n' % (stmt.keyword, stmt.arg))
        if name in stmt.i_typedefs:
            w('  FOUND\n')
            t = stmt.i_typedefs[name]
            if (mod is not None and
                mod != t.i_orig_module and
                t.i_orig_module.keyword == 'submodule'):
                # make sure this submodule is included
                if mod.search_one('include', t.i_orig_module.arg) is None:
                    return None
            return t
        stmt = stmt.parent
    return None
```
```

wlupton commented:
@mbj4668 I was just wondering whether you had had a chance to look at this.

mbj4668 commented:
> @mbj4668 I was just wondering whether you had had a chance to look at this.

No. I will try to find some time in the coming week.

wlupton commented:
The recent updates were due to a rebase against master. There were conflicts but I hope that I resolved them all in favour of master. Only these files on this branch are different from master:
```
% git diff --name-only master
pyang/transforms/edit.py
test/test_transform/add-foo.py
test/test_transform/out/a--transform=add-foo--format=tree.out
test/test_transform/out/a--transform=add-foo--transform=mod-desc--format=yang.out
test/test_transform/out/a--transform=mod-desc--format=yang.out
test/test_transform/out/a--transform=mod-desc--transform=add-foo--format=yang.out
test/test_transform/out/a.out
```

wlupton commented:
@mbj4668, using the earlier-presented `test.yang` I ran this in this branch (see the bottom of this message for the `null` transform):
```
% pyang test.yang
<no output>
% pyang -t null test.yang 
test.yang:4: warning: imported module "ietf-interfaces" not used
```

I then cherry-picked the changes from `mbj/transform-reset-env` into this branch and the warning went away.

I also tried this version (with the real `edit` transform) on some real YANG and it looks OK to me, i.e the warnings went away and I didn't see anything suggesting problems with the statement hierarchy.

So I'm wondering whether this PR can finally be merged? Possible reasons not to:
* There's a Travis flake8 error complaining about a "\d" escape; I'll look into that and fix it
* There's the question of whether the `edit` transform should go into `pyang/transforms` or into an `examples` directory (ideally with a `--search-examples`, or similar, option that will automatically search it)

Your thoughts?

This is the `null` transform:
```
% cat ./pyang/transforms/null.py
from pyang import plugin

def pyang_plugin_init():
    plugin.register_plugin(Null())

class Null(plugin.PyangPlugin):
    def add_transform(self, xforms):
        xforms['null'] = self

    def transform(self, ctx, modules):
	pass
```

wlupton commented:
@mbj4668, I'm not sure why the build is still failing (git ls-files is generating unexpected output). Could this be because the PR adds some new test files that Travis isn't considering?

wlupton commented:
I've rebased against the latest master. The checks still have the same failure, which is possibly because the PR adds some new test files that Travis isn't considering? 
```
$ git ls-files -i --exclude-standard | tee "$TMP"
test/test_transform/out/a--transform=add-foo--format=tree.out
test/test_transform/out/a--transform=add-foo--transform=mod-desc--format=yang.out
test/test_transform/out/a--transform=mod-desc--format=yang.out
test/test_transform/out/a--transform=mod-desc--transform=add-foo--format=yang.out
test/test_transform/out/a.out
The command "git ls-files -i --exclude-standard | tee "$TMP"" exited with 0.
$ if test -s "$TMP"; then false; else true; fi
The command "if test -s "$TMP"; then false; else true; fi" exited with 1.
```

mbj4668 commented:
The issue is that you have checked in the generated files in out/.   Remove the "out" dir from git.

wlupton commented:
Oops, sorry. Silly me. I've done this and hope that the tests will now pass.
